### PR TITLE
Added XML reporting #430

### PR DESCRIPTION
**Pull request by jacarandang**
I tried working on XML report. I'm really new to brakeman and XML, any comments would be highly appreciated.
#430

### CODE REVIEW

**Comment by presidentbeef** [comment]
Hi,

Thanks for looking into this!

`:xml` should be added as an option here: https://github.com/presidentbeef/brakeman/blob/master/lib/brakeman/options.rb#L166-L169

Unfortunately, your branch isn't working for me on any apps I've tested. This is the error:

```
/home/justin/.rvm/gems/ruby-2.1.5@brakeman/gems/builder-3.2.2/lib/builder/xmlbase.rb:61:in `tag!': XmlMarkup cannot mix a text argument with a block (ArgumentError)
	from /home/justin/.rvm/gems/ruby-2.1.5@brakeman/gems/builder-3.2.2/lib/builder/xmlbase.rb:93:in `method_missing'
	from /home/justin/work/brakeman/lib/brakeman/report/report_xml.rb:48:in `output_xml'
	from /home/justin/work/brakeman/lib/brakeman/report/report_xml.rb:33:in `generate_report'
	from /home/justin/work/brakeman/lib/brakeman/report.rb:76:in `generate'
	from /home/justin/work/brakeman/lib/brakeman/report.rb:45:in `format'
	from /home/justin/work/brakeman/lib/brakeman.rb:356:in `block (2 levels) in write_report_to_files'
	from /home/justin/work/brakeman/lib/brakeman.rb:355:in `open'
	from /home/justin/work/brakeman/lib/brakeman.rb:355:in `block in write_report_to_files'
	from /home/justin/work/brakeman/lib/brakeman.rb:354:in `each'
	from /home/justin/work/brakeman/lib/brakeman.rb:354:in `each_with_index'
	from /home/justin/work/brakeman/lib/brakeman.rb:354:in `write_report_to_files'
	from /home/justin/work/brakeman/lib/brakeman.rb:343:in `scan'
	from /home/justin/work/brakeman/lib/brakeman.rb:62:in `run'
	from /home/justin/work/brakeman/bin/brakeman:78:in `<main>'
```

**Comment by jacarandang** [comment]
Thank you for checking my PR. Can I see the text output? I think it is caused by errors.

**Comment by jacarandang** [comment]
As expected. the errors was caused by the error reports. Can you try again now? I'll add tests later. Thank you

**Comment by presidentbeef** [comment]
I do not think errors are the problem. Pick any of the apps in https://github.com/presidentbeef/brakeman/tree/master/test/apps (none of which should have errors) and you will still see the issue.

**Comment by jacarandang** [comment]
Weird, I tried it against the apps and it did work well. I'll check on this, I'll try to make tests. Thanks

**Comment by presidentbeef** [comment]
I'm using Ruby 2.1.5 with Builder 3.2.2, if that helps.

**Comment by jacarandang** [comment]
I think the send method is causing the problem. It might also be overwritten by Builder, I used the tag! method instead and added sanity check.

**Comment by jacarandang** [comment]
Builder doesn't seem to support ruby 1.8.7 which causes the travis to fail.

File: lib/brakeman/report/report_xml.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    output_xml x, :report_info, reportInfo
+
+  end
+
+  def convert_to_hashes warnings, typ
+    warnings.map do |w|
+      hash = w.to_hash
+      hash[:file] = warning_file w
+      hash[:location] = hash[:location].flatten.join(',') unless hash[:location].nil?
```
Comment (presidentbeef) [inline lib/brakeman/report/report_xml.rb]: Shouldn't this be nice XML, not just a string?

File: lib/brakeman/report/report_xml.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    scanInfo = {
+      :app_path => File.expand_path(tracker.app_path),
+      :ruby_version => RUBY_VERSION,
+      :rails_version => rails_version,
+      :brakeman_version => Brakeman::Version,
+      :report_generation_time => Time.now.to_s,
+      :start_time => tracker.start_time.to_s,
+      :end_time => tracker.end_time.to_s,
+      :duration => tracker.duration,
+      :checks => checks.checks_run.sort.join(',')
```
Comment (presidentbeef) [inline lib/brakeman/report/report_xml.rb]: Shouldn't this be a list in XML, not a string?

Comment (jacarandang) [inline lib/brakeman/report/report_xml.rb]: What do you mean? Is this the checks?

File: lib/brakeman/report/report_xml.rb (hunk 1 above)
Comment (jacarandang) [inline lib/brakeman/report/report_xml.rb]: Is this about the location?

File: lib/brakeman/report/report_xml.rb (hunk 2 above)
Comment (presidentbeef) [inline lib/brakeman/report/report_xml.rb]: Yes. Currently it is turning an array into a string.

File: lib/brakeman/report/report_xml.rb (hunk 1 above)
Comment (presidentbeef) [inline lib/brakeman/report/report_xml.rb]: Yes.

**Comment by presidentbeef** [comment]
It would be nice if the test could parse the XML back into a hash and compare it with the expected values.

File: brakeman-min.gemspec [hunk 3]
```diff
@@ -18,4 +18,5 @@ Gem::Specification.new do |s|
   s.add_dependency "ruby_parser", "~>3.7.0"
   s.add_dependency "ruby2ruby", "~>2.1.1"
   s.add_dependency "multi_json", "~>1.2"
+  s.add_dependency "builder", "~>2.0"
```
Comment (presidentbeef) [inline brakeman-min.gemspec]: Don't add builder here, brakeman-min only supports JSON by default.

File: lib/brakeman/report/report_xml.rb [hunk 4]
```diff
@@ -0,0 +1,78 @@
+require 'builder'
```
Comment (presidentbeef) [inline lib/brakeman/report/report_xml.rb]: Please use `Brakeman.load_brakeman_dependency 'builder'` instead of `require`

**Comment by presidentbeef** [comment]
I'm not sure we should use `builder` given it seems unmaintained since Jim Weirich passed away, but it's probably okay.

**Comment by jacarandang** [comment]
How about using Nokogiri instead? Could also help in parsing XML files for tests.

**Comment by presidentbeef** [comment]
That would be fine with me.

**Comment by dvandersluis** [comment]
FYI - I've been doing a bunch of testing of XML builders lately, and the `builder` gem is quite slow. `nokogiri` is significantly faster, and in my testing [`ox`](https://github.com/ohler55/ox) is even better (it's recently been updated to have a module for building XML documents).

**Comment by presidentbeef** [comment]
I'm open to another attempt at this with a maintained library.

**Pull request closed without merging**